src/ports/postgres/modules/knn/knn.py_in

```diff
@@ -124,16 +137,312 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
+                                        data points.
+            @param test_column_name     Name of the column with testing data
+                                        points or expression that evaluates to a
+                                        numeric array
+            @param test_id              Name of the column having ids of data
+                                        points in test data table.
+            @param interim_table        Name of the table to store interim
+                                        results.
+            @param in_k                 default: 1. Number of nearest
+                                        neighbors to consider
+            @param output_neighbours    Outputs the list of k-nearest neighbors
+                                        that were used in the voting/averaging.
+            @param fn_dist              Distance metrics function. Default is
+                                        squared_dist_norm2. Following functions
+                                        are supported :
+                                        dist_norm1 , dist_norm2,squared_dist_norm2,
+                                        dist_angle , dist_tanimoto
+                                        Or user defined function with signature
+                                        DOUBLE PRECISION[] x, DOUBLE PRECISION[] y -> DOUBLE PRECISION
+            @param weighted_avg         Calculates the Regression or classication of k-NN using
+                                        the weighted average method.
+            @param leaf_nodes           Number of leaf nodes to explore
+            @param r_id                 Name of the region id column
+            @param dim                  Name of the dimension column
+            Following parameters are passed to ensure the interim table has
+            identical features to non-kd-tree implementation
+            @param label_out
+            @param comma_label_out_alias
+            @param label_name
+            @param train
+            @param train_id
+            @param dist_inverse
+            @param test_id_temp
+    """
+    with MinWarning("error"):
+
+        tree_model = add_postfix(kd_out, "_tree")
+        centers_table = add_postfix(kd_out, "_centers")
+        n_features = num_features(test_source, test_column_name)
+
+        tree = plpy.execute("SELECT * FROM {0}".format(tree_model))[0]['tree']
+        # 'tree' contains only non-leaf nodes,
+        # hence 'n_leaves' is always 1 more than len(tree)
+        n_leaves = len(tree)+1
+
+        depth = int(log(n_leaves, 2))
+
+        # The borders table will have two rows for each dimension (Upper & lower)
+        # even if a dimension does not have a branch.
+        # Its borders will be -Inf, Inf
+
+        # The first leaf_note is itself,
+        # we expand to n-1 nodes out of 2 * n_features borders
+        quant = float(leaf_nodes) / n_leaves
+
+        clause_counter = 0
```

Comment:
The logic here is the same as the one in tree building. We can either abstract to a separate function or just store the complete string as the tree to avoid rebuilding here.